k2/csrc/thread_pool.h

```diff
@@ -0,0 +1,108 @@
+/**
+ * @brief Thread pool.
+ *
+ *
+ * @copyright
+ * Copyright (c)  2020  Xiaomi Corporation (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+#ifndef K2_CSRC_THREAD_POOL_H_
+#define K2_CSRC_THREAD_POOL_H_
+
+#include <condition_variable>  // NOLINT
+#include <functional>
+#include <mutex>  // NOLINT
+#include <queue>
+#include <thread>  // NOLINT
+#include <vector>
+
+#include "k2/csrc/log.h"
+
+namespace k2 {
+
+class ThreadPool {
+ public:
+  using TaskFunc = std::function<void()>;
+
+  /* Create a pool with given number of threads.
+   *
+   * @param [in] num_threads  The number of threads in the pool. If it is <= 0,
+   *                          the number of threads in the pool is set to
+   *                          `std::thread::hardware_concurrency()`.
+   */
+  explicit ThreadPool(int32_t num_threads);
+
+  /* It ask all threads in the pool to exit after processing what they
+   * are currently executing. If there are still tasks in the queue, they
+   * will not be executed.
+   */
+  ~ThreadPool();
+
+  /* Insert a task into the queue. It is executed immediately
+   * if there are currently idle threads. Otherwise, it is saved
+   * in the queue and waits for threads to be free to pop it from
+   * the queue and to execute it.
+   *
+   * @param [in] task   The task to be run. It should be convertible to
+   *                    `std::function<void()>`.
+   */
+  template <typename Lambda>
+  void RunTask(Lambda task) {
+    K2_CHECK_GT(threads_.size(), 0u);
+    std::lock_guard<std::mutex> lock(mutex_);
+    tasks_.emplace(static_cast<TaskFunc>(task));
+    not_empty_cond_.notify_one();
+    finished_ = false;
+  }
+
+  /* The caller is blocked if `tasks_` is not empty.
+   * If `tasks_` is empty, it returns immediately.
+   */
+  void WaitAllTasksFinished();
+
+ private:
+  /* Remove tasks from `task_` and process them.
```

Comment:
I would say something like, this function keep asking (in an infinite loop) if there's any un-executed task in `task_`, if there's any one, it will remove it from `task_` and process it. If `keep_running_` is false, it will return after processing current task